pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/JSONIndexHandler.java

```diff
@@ -0,0 +1,216 @@
+      FileUtils.touch(inProgress);
+    } else {
+      // Marker file exists, which means last run gets interrupted.
+
+      // Remove json index if exists.
+      // For v1 and v2, it's the actual json index. For v3, it's the temporary json index.
+      FileUtils.deleteQuietly(jsonIndexFile);
+    }
+
+    // Create new json index for the column.
+    LOGGER.info("Creating new json index for segment: {}, column: {}", _segmentName, column);
+    if (columnMetadata.hasDictionary()) {
+      handleDictionaryBasedColumn(columnMetadata);
+    } else {
+      handleNonDictionaryBasedColumn(columnMetadata);
+    }
+
+    // For v3, write the generated json index file into the single file and remove it.
+    if (_segmentVersion == SegmentVersion.v3) {
+      LoaderUtils.writeIndexToV3Format(_segmentWriter, column, jsonIndexFile, ColumnIndexType.JSON_INDEX);
+    }
+
+    // Delete the marker file.
+    FileUtils.deleteQuietly(inProgress);
+
+    LOGGER.info("Created json index for segment: {}, column: {}", _segmentName, column);
+  }
+
+  private void handleDictionaryBasedColumn(ColumnMetadata columnMetadata)
+      throws IOException {
+    int numDocs = columnMetadata.getTotalDocs();
+    try (ForwardIndexReader forwardIndexReader = getForwardIndexReader(columnMetadata, _segmentWriter);
+        ForwardIndexReaderContext readerContext = forwardIndexReader.createContext();
+        Dictionary dictionary = getDictionaryReader(columnMetadata, _segmentWriter);
+        JSONIndexCreator jsonIndexCreator = new JSONIndexCreator(_indexDir, columnMetadata.getFieldSpec())) {
+      if (columnMetadata.isSingleValue()) {
+        switch (columnMetadata.getDataType()) {
+          case STRING:
+            for (int i = 0; i < numDocs; i++) {
+              int dictId = forwardIndexReader.getDictId(i, readerContext);
+              jsonIndexCreator.add(dictionary.getStringValue(dictId).getBytes("UTF-8"));
+            }
+            break;
+          case BYTES:
+            // Single-value column
+            for (int i = 0; i < numDocs; i++) {
+              int dictId = forwardIndexReader.getDictId(i, readerContext);
+              jsonIndexCreator.add(dictionary.getBytesValue(dictId));
+            }
+            break;
+          default:
+            throw new IllegalStateException("Unsupported data type: " + columnMetadata.getDataType());
+        }
+      } else {
+        // Multi-value column
+        throw new IllegalStateException("JSON Indexing is not supported on multi-valued columns ");
+      }
+      jsonIndexCreator.seal();
+    }
+  }
+
+  private void handleNonDictionaryBasedColumn(ColumnMetadata columnMetadata)
+      throws IOException {
+    FieldSpec.DataType dataType = columnMetadata.getDataType();
+    if (dataType != FieldSpec.DataType.BYTES || dataType != FieldSpec.DataType.STRING) {
+      throw new UnsupportedOperationException(
+          "JSON indexing is only supported for STRING/BYTES datatype but found: " + dataType);
```

Comment:
why do not support number type?

Reply:
+1.

Reply:
Json field must be of `STRING` type, and the value should be json strings